internal/processor/stream/processor.go

```diff
@@ -212,43 +218,60 @@ func (p *Processor) readBatch(
 // Callers must not access result concurrently with HandleStream.
 func (p *Processor) HandleStream(
 	ctx context.Context,
+	async bool,
 	baseEvent model.APMEvent,
 	reader io.Reader,
 	batchSize int,
-	processor model.BatchProcessor,
+	proc model.BatchProcessor,
 	result *Result,
 ) error {
-	// Since processor.ProcessBatch can block for some time until all the batch
-	// is added to the modelindexer's cache and there isn't a fail-fast rejection,
-	// we want to cap how many in-flight requests are read at any time.
+	// Limits the number of concurrent batch decodes.
 	// Defaults to 200 (N), only allowing N requests to read and cache Y events
-	// (determined by batchSize) from the batch, effectively limitting the decoding
-	// concurrency to N batches at any time, which should be a good ceiling. The
-	// ceiling also reduces the contention on the modelindexer.activeMu.
-	select {
-	case p.sem <- struct{}{}:
-	case <-ctx.Done():
-		return ctx.Err()
+	// (determined by batchSize) from the batch.
+	// The ceiling may also reduce the contention on the modelindexer.activeMu.
+	// Clients can set a async to true which makes the processor handle the
+	// events in the background. Returns with an error `publish.ErrFull` if the
+	// semaphore is full.
+	if err := p.acquireLock(ctx, async); err != nil {
+		return err
 	}
-
 	sr := p.getStreamReader(reader)
-	defer func() {
-		sr.release()
-		<-p.sem
-	}()
 
 	// first item is the metadata object
 	if err := p.readMetadata(sr, &baseEvent); err != nil {
 		// no point in continuing if we couldn't read the metadata
+		sr.release()
+		p.releaseLock()
 		return err
 	}
 
+	defer func() {
+		sr.release()
+		if !async { // Release lock for sync processing.
+			p.releaseLock()
+		}
+	}()
+
 	sp, ctx := apm.StartSpan(ctx, "Stream", "Reporter")
 	defer sp.End()
 
+	// Create a new processor var to avoid replacing the passed reference.
+	var processor model.BatchProcessor
+	if async {
+		// Allow 5 async batches to be processing per stream. The lambda
+		// extension doesn't (yet) stream to the server.
+		// If a single batch with more than the batchSize is sent (10),
+		// then the request can process 50 events asynchronously.
+		batchChan := make(chan *model.Batch, 5)
```

Comment:
Hmm, now that I see this in code: it means we could have higher memory usage for async=true, right? i.e. if you had 200 concurrent requests. Also this channel size feels a bit arbitrary. (I realise this is what I described in my other comment, I just hadn't quite thought it through.)

Would it make sense to insead allow a single request to process up to 200 (i.e. semaphore size) batches asynchronously? Something like this, in the loop below:
1. acquire the semaphore at the top of the loop, returning 503 if there's none available
2. call readBatch
3. start a new goroutine to process the batch; the goroutine will release semaphore when processing is done